my-megaparsec-project/src/Main.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Main (main) where
+
+import Data.Text (Text)
+import qualified Data.Text as T
+import Data.Void
+import Text.Megaparsec
+import Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer as L
+
+type Parser = Parsec Void Text
+
+data Expr
+  = IntLit Int
+  | Sum Expr Expr
+  | Sub Expr Expr
+  | Prod Expr Expr
+  | Div Expr Expr
+  | Neg Expr
+  deriving (Eq, Show)
+
+evalExpr :: Expr -> Int
+evalExpr (IntLit n) = n
+evalExpr (Sum a b) = evalExpr a + evalExpr b
+evalExpr (Sub a b) = evalExpr a - evalExpr b
+evalExpr (Prod a b) = evalExpr a * evalExpr b
+evalExpr (Div a b) = evalExpr a `div` evalExpr b
+evalExpr (Neg a) = -(evalExpr a)
+
+prettyPrintExpr :: Expr -> String
+prettyPrintExpr (IntLit n) = show n
+prettyPrintExpr (Sum a b)  = prettyPrintExpr a ++ " + " ++ prettyPrintExpr b
+prettyPrintExpr (Sub a b)  = prettyPrintExpr a ++ " - " ++ prettyPrintParen b
+prettyPrintExpr (Prod a b) = prettyPrintParen a ++ " * " ++ prettyPrintParen b
+prettyPrintExpr (Div a b)  = prettyPrintParen a ++ " / " ++ prettyPrintParen b
+prettyPrintExpr (Neg a) = "-" ++ prettyPrintExpr a
+
+prettyPrintParen :: Expr -> String
+prettyPrintParen e = case e of
+  IntLit _ -> prettyPrintExpr e
+  _        -> "(" ++ prettyPrintExpr e ++ ")"
+
+spaceConsumer :: Parser ()
+spaceConsumer = L.space space1 empty empty
+
+symbol :: Char -> Parser Char
+symbol c = char c <* spaceConsumer
+
+parseInt :: Parser Expr
+parseInt = do
+  n <- L.decimal <* spaceConsumer
+  return (IntLit n)
+
+parseNegative :: Parser Expr
+parseNegative = do
+  _ <- symbol '-'
+  Neg <$> (parseInt <|> parseExpr)
+
+parseParens :: Parser Expr
+parseParens = do
+  _ <- symbol '('
+  expr <- parseExpr
+  _ <- symbol ')'
+  return expr
+
+parseFactor :: Parser Expr
+parseFactor = parseInt <|> parseNegative <|> parseParens
+
+-- parseTerm = chain of factors, split by * or /
+parseTerm :: Parser Expr
+parseTerm = do
+  first <- parseFactor
+  rest <- many $ do
+    op <- (symbol '*' >> return Prod) <|> (symbol '/' >> return Div)
+    expr <- parseFactor
+    return (op, expr)
+  return $ foldl (\acc (op, x) -> op acc x) first rest
+
+-- parseExpr = cahin of Terms, split + by -
+parseExpr :: Parser Expr
+parseExpr = do
+  first <- parseTerm
+  rest <- many $ do
+    op <- (symbol '+' >> return Sum) <|> (symbol '-' >> return Sub)
+    expr <- parseTerm
+    return (op, expr)
+  return $ foldl (\acc (op, x) -> op acc x) first rest
+
```

Comment:
This will do the work just fine, but I'd recommend using an "uber"-expression parser constructed from a table. See the details [here](https://markkarpov.com/tutorial/megaparsec.html#parsing-expressions)